src/hotspot/os/linux/os_linux.cpp

```diff
@@ -5872,30 +5874,83 @@ static int cr_util_path(char* path, int len) {
 }
 
 static bool compute_crengine() {
+  // release possible old copies
+  os::free((char *) _crengine); // NULL is allowed
+  _crengine = NULL;
+  os::free((char *) _crengine_arg_str);
+  _crengine_arg_str = NULL;
+
   if (!CREngine) {
     return true;
   }
-
-  if (CREngine[0] == '/') {
-    _crengine = CREngine;
-    return true;
+  char *exec = os::strdup_check_oom(CREngine);
+  char *comma = strchr(exec, ',');
+  if (comma != NULL) {
+    *comma = '\0';
+    _crengine_arg_str = os::strdup_check_oom(comma + 1);
   }
-
-  char path[JVM_MAXPATHLEN];
-  int pathlen = cr_util_path(path, sizeof(path));
-  strcat(path + pathlen, "/");
-  strcat(path + pathlen, CREngine);
-
-  struct stat st;
-  if (0 != stat(path, &st)) {
-    warning("Could not find %s: %s", path, strerror(errno));
-    return false;
+  if (exec[0] == '/') {
+    _crengine = exec;
+  } else {
+    char path[JVM_MAXPATHLEN];
+    int pathlen = cr_util_path(path, sizeof(path));
+    strcat(path + pathlen, "/");
+    strcat(path + pathlen, exec);
+
+    struct stat st;
+    if (0 != stat(path, &st)) {
+      warning("Could not find %s: %s", path, strerror(errno));
+      return false;
+    }
+    _crengine = os::strdup_check_oom(path);
+    // we have read and duplicated args from exec, now we can release
+    os::free(exec);
   }
+  _crengine_args[0] = _crengine;
 
-  _crengine = os::strdup(path);
+  size_t next_arg = 2;
+  if (_crengine_arg_str != NULL) {
+    char *arg = _crengine_arg_str;
+    char *target = _crengine_arg_str;
+    bool escaped = false;
+    for (char *c = arg; *c != '\0'; ++c) {
+      if (next_arg >= ARRAY_SIZE(_crengine_args) - 2) {
+        warning("Too many options to CREngine; cannot proceed with these: %s", arg);
+        return false;
+      }
+      if (!escaped) {
+        switch(*c) {
+        case '\\':
+          escaped = true;
+          continue; // for
+        case ',':
+          *target++ = '\0';
+          _crengine_args[next_arg++] = arg;
+          arg = target;
+          continue; // for
+        }
+      }
+      escaped = false;
+      *target++ = *c;
+    }
+    *target = '\0';
+    _crengine_args[next_arg++] = arg;
+    _crengine_args[next_arg] = NULL;
+  }
   return true;
 }
 
+static void add_crengine_arg(const char *arg) {
+  for (size_t i = 2; i < ARRAY_SIZE(_crengine_args) - 1; ++i) {
+    if (_crengine_args[i] == NULL) {
```

Comment:
The last empty position can be remembered along _crengine_args

Reply:
The fewer static vars the better IMO, but I can certainly use an explicit index var and drop the initialization to three NULLs if you think it's better.

Reply:
In general I agree, but since _crengine_args is already stateful, I don't see any troubles adding more statics that is in fact a part of the same state. But this is minor, so up to you. OK for me as it is.